File: android/libs/googleassistant/src/main/java/app/organicmaps/intent/geo/GeoIntentProcessor.java
package app.organicmaps.intent.geo;

import android.content.Intent;
import android.text.TextUtils;
import androidx.annotation.NonNull;
import app.organicmaps.intent.IntentProcessor;
import app.organicmaps.intent.geo.navigation.GeoIntent;
import app.organicmaps.sdk.util.log.Logger;

public class GeoIntentProcessor
{
  private static final String TAG = GeoIntentProcessor.class.getSimpleName();

  private final IntentProcessor mIntentProcessor;

  public GeoIntentProcessor(@NonNull IntentProcessor intentProcessor)
  {
    mIntentProcessor = intentProcessor;
  }

  public boolean processIntent(@NonNull Intent intent)
  {
    if (GeoNavigationIntent.isNavigationIntent(intent))
      processNavigationIntent(intent);
    else if (GeoViewIntent.isSearchIntent(intent))
      processSearchIntent(intent);
    else if (GeoActionIntent.isActionIntent(intent))
      processActionIntent(intent);
    else
      return false;
    return true;
  }

  private void processNavigationIntent(@NonNull Intent intent)
  {
    final GeoNavigationIntent navigationIntent = GeoNavigationIntent.fromIntent(intent);
    final double lat = navigationIntent.getLat();
    final double lon = navigationIntent.getLon();
    final String query = navigationIntent.getQuery();
    final boolean areCoordinatesValid = !Double.isNaN(lat);
    final boolean hasQuery = !TextUtils.isEmpty(query);
    final GeoIntent geoIntent = navigationIntent.getIntent();
    if (geoIntent == GeoIntent.Navigation)
    {
      if (areCoordinatesValid && hasQuery)
        mIntentProcessor.navigate(lat, lon, query, navigationIntent.getAvoidList(), navigationIntent.getTravelMode());
      else if (areCoordinatesValid)
        mIntentProcessor.navigate(lat, lon, navigationIntent.getAvoidList(), navigationIntent.getTravelMode());
      else if (hasQuery)
        mIntentProcessor.navigate(query, navigationIntent.getAvoidList(), navigationIntent.getTravelMode());
    }
    else if (geoIntent == GeoIntent.Directions)
    {
      if (areCoordinatesValid && hasQuery)
        mIntentProcessor.showDirections(lat, lon, query, navigationIntent.getAvoidList(),
                                        navigationIntent.getTravelMode());
      else if (areCoordinatesValid)
        mIntentProcessor.showDirections(lat, lon, navigationIntent.getAvoidList(), navigationIntent.getTravelMode());
      else if (hasQuery)
        mIntentProcessor.showDirections(query, navigationIntent.getAvoidList(), navigationIntent.getTravelMode());
    }
    else if (geoIntent == GeoIntent.AddStop)
    {
      if (areCoordinatesValid && hasQuery)
        mIntentProcessor.addStop(lat, lon, query);
      else if (areCoordinatesValid)
        mIntentProcessor.addStop(lat, lon);
      else if (hasQuery)
        mIntentProcessor.addStop(query);
    }
  }

  private void processSearchIntent(@NonNull Intent intent)
  {
    final GeoViewIntent searchIntent = GeoViewIntent.fromIntent(intent);
    final String query = searchIntent.getQuery();
    final double lat = searchIntent.getLat();
    final double lon = searchIntent.getLon();
    if (!TextUtils.isEmpty(query))
    {
      if (!Double.isNaN(lat))
        mIntentProcessor.search(lat, lon, query);
      else
        mIntentProcessor.search(query);
    }
  }

  private void processActionIntent(@NonNull Intent intent)
  {
    final GeoActionIntent actionIntent = GeoActionIntent.fromIntent(intent);
    switch (actionIntent.getActionType())
    {
    case AllowFerries -> mIntentProcessor.processFerriesAction(true);
    case AllowHighways -> mIntentProcessor.processHighwaysAction(true);
    case AllowTolls -> mIntentProcessor.processTollsAction(true);
    case AvoidFerries -> mIntentProcessor.processFerriesAction(false);
    case AvoidHighways -> mIntentProcessor.processHighwaysAction(false);
    case AvoidTolls -> mIntentProcessor.processTollsAction(false);
    case DistanceToDestination -> mIntentProcessor.speakDistanceToDestination();
    case DistanceToNextTurn -> mIntentProcessor.speakDistanceToNextTurn();
    case Eta -> mIntentProcessor.speakEta();
    case ExitNavigation -> mIntentProcessor.exitNavigation();
    case FollowMode -> mIntentProcessor.changeMapViewToFollowMode();
    case GoBack -> mIntentProcessor.goBack();
    case HideSatellite -> mIntentProcessor.processShowSatelliteAction(false);
    case HideTraffic -> mIntentProcessor.processShowTrafficAction(false);
    case Mute -> mIntentProcessor.processVoiceGuidanceAction(false);
    case QueryCurrentRoad -> mIntentProcessor.speakCurrentRoad();
    case QueryDestination -> mIntentProcessor.speakDestination();
    case QueryNextTurn -> mIntentProcessor.speakNextTurn();
    case ReportCrash -> mIntentProcessor.reportCrash(actionIntent.getAccidentType(), actionIntent.getRoadDirection());
    case ReportHazard ->
      mIntentProcessor.reportHazard(actionIntent.getHazardType(), actionIntent.getLocationOnRoad(),
                                    actionIntent.getRoadDirection());
    case ReportPolice -> mIntentProcessor.reportPolice(actionIntent.getRoadDirection());
    case ReportRoadClosure -> mIntentProcessor.reportRoadClosure(actionIntent.getRoadClosureType());
    case ReportTraffic ->
      mIntentProcessor.reportTraffic(actionIntent.getTrafficType(), actionIntent.getRoadDirection());
    case RouteOverview -> mIntentProcessor.showRouteOverview();
    case ShowAlternativeRoutes, ShowAlternates -> mIntentProcessor.showAlternativeRoutes();
    case ShowDirectionsList -> mIntentProcessor.showDirectionsList();
    case ShowSatellite -> mIntentProcessor.processShowSatelliteAction(true);
    case ShowTraffic -> mIntentProcessor.processShowTrafficAction(true);
    case TimeToDestination -> mIntentProcessor.speakTimeToDestination();
    case TimeToNextTurn -> mIntentProcessor.speakTimeToNextTurn();
    case Unmute -> mIntentProcessor.processVoiceGuidanceAction(true);
    case MyLocation -> mIntentProcessor.showMyLocation();
    case ShowMap -> mIntentProcessor.showMap();
    case TrafficReport -> mIntentProcessor.speakTrafficReport();
    case ClearSearchResults -> mIntentProcessor.clearSearchResults();
    case SelectSearchResult -> mIntentProcessor.selectSearchResult(actionIntent.getSearchId());
    case ApplyElectricVehicleConnectorFilter -> mIntentProcessor.enableElectricVehicleConnectorFilter(true);
    case RemoveElectricVehicleConnectorFilter -> mIntentProcessor.enableElectricVehicleConnectorFilter(false);
    case ApplyElectricVehiclePaymentFilter -> mIntentProcessor.enableElectricVehiclePaymentFilter(true);
    case RemoveElectricVehiclePaymentFilter -> mIntentProcessor.enableElectricVehiclePaymentFilter(false);
    case ApplyElectricVehicleFastChargingFilter -> mIntentProcessor.enableElectricVehicleFastChargingFilter(true);
    case RemoveElectricVehicleFastChargingFilter -> mIntentProcessor.enableElectricVehicleFastChargingFilter(false);
    default -> Logger.w(TAG, "Unknown action type: " + actionIntent.getActionType());
    }
  }
}

File: android/libs/googleassistant/src/main/java/app/organicmaps/intent/geo/GeoIntentProcessor.kt
package app.organicmaps.intent.geo

import android.content.Intent
import app.organicmaps.intent.geo.action.ActionType
import app.organicmaps.intent.geo.handlers.AvoidActionHandler
import app.organicmaps.intent.geo.handlers.ControlActionHandler
import app.organicmaps.intent.geo.handlers.NavigationActionHandler
import app.organicmaps.intent.geo.handlers.ReportActionHandler
import app.organicmaps.intent.geo.handlers.SearchActionHandler
import app.organicmaps.intent.geo.handlers.VoiceActionHandler
import app.organicmaps.intent.geo.navigation.UserIntent
import app.organicmaps.sdk.util.log.Logger

class GeoIntentProcessor(
    private val mAvoidActionHandler: AvoidActionHandler,
    private val mControlActionHandler: ControlActionHandler,
    private val mNavigationActionHandler: NavigationActionHandler,
    private val mReportActionHandler: ReportActionHandler,
    private val mSearchActionHandler: SearchActionHandler,
    private val mVoiceActionHandler: VoiceActionHandler,
) {
    fun processIntent(intent: Intent): Boolean {
        val navigationIntent = GeoNavigationIntent.fromIntent(intent)
        if (navigationIntent != null) {
            processNavigationIntent(navigationIntent)
            return true
        }

        val viewIntent = GeoViewIntent.fromIntent(intent)
        if (viewIntent != null) {
            processViewIntent(viewIntent)
            return true
        }

        val actionIntent = GeoActionIntent.fromIntent(intent)
        if (actionIntent != null) {
            processActionIntent(actionIntent)
            return true
        }

        return false
    }

    private fun processNavigationIntent(navigationIntent: GeoNavigationIntent) {
        val lat = navigationIntent.lat
        val lon = navigationIntent.lon
        val query = navigationIntent.query
        val geoIntent = navigationIntent.intent
        when (geoIntent) {
            UserIntent.Navigation -> mNavigationActionHandler.navigate(
                lat,
                lon,
                query,
                navigationIntent.avoidList,
                navigationIntent.travelMode,
            )

            UserIntent.Directions -> mNavigationActionHandler.showDirections(
                lat,
                lon,
                query,
                navigationIntent.avoidList,
                navigationIntent.travelMode,
            )

            UserIntent.AddStop -> mNavigationActionHandler.addStop(lat, lon, query)
        }
    }

    private fun processViewIntent(viewIntent: GeoViewIntent) {
        val query = viewIntent.query
        val lat = viewIntent.lat
        val lon = viewIntent.lon

        mSearchActionHandler.search(lat, lon, query)
    }

    private fun processActionIntent(actionIntent: GeoActionIntent) {
        when (val actionType = actionIntent.actionType) {
            in AVOID_ACTIONS -> processAvoidAction(actionType)
            in CONTROL_ACTIONS -> processControlAction(actionType)
            in VOICE_ACTIONS -> processVoiceAction(actionType)
            in REPORT_ACTIONS -> processReportAction(actionIntent)
            in SEARCH_ACTIONS -> processSearchAction(actionIntent)
            else -> Logger.w(TAG, "Unhandled action type: $actionType")
        }
    }

    private fun processAvoidAction(actionType: ActionType) {
        when (actionType) {
            ActionType.AllowFerries -> mAvoidActionHandler.processFerriesAction(true)
            ActionType.AllowHighways -> mAvoidActionHandler.processHighwaysAction(true)
            ActionType.AllowTolls -> mAvoidActionHandler.processTollsAction(true)
            ActionType.AvoidFerries -> mAvoidActionHandler.processFerriesAction(false)
            ActionType.AvoidHighways -> mAvoidActionHandler.processHighwaysAction(false)
            ActionType.AvoidTolls -> mAvoidActionHandler.processTollsAction(false)
            else -> Unit // Unreachable: caller only dispatches AVOID_ACTIONS here.
        }
    }

    private fun processControlAction(actionType: ActionType) {
        when (actionType) {
            ActionType.ExitNavigation -> mControlActionHandler.exitNavigation()
            ActionType.FollowMode -> mControlActionHandler.changeMapViewToFollowMode()
            ActionType.GoBack -> mControlActionHandler.goBack()
            ActionType.HideSatellite -> mControlActionHandler.processShowSatelliteAction(false)
            ActionType.HideTraffic -> mControlActionHandler.processShowTrafficAction(false)
            ActionType.RouteOverview -> mControlActionHandler.showRouteOverview()
            ActionType.ShowAlternativeRoutes, ActionType.ShowAlternates -> mControlActionHandler.showAlternativeRoutes()
            ActionType.ShowDirectionsList -> mControlActionHandler.showDirectionsList()
            ActionType.ShowSatellite -> mControlActionHandler.processShowSatelliteAction(true)
            ActionType.ShowTraffic -> mControlActionHandler.processShowTrafficAction(true)
            ActionType.MyLocation -> mControlActionHandler.showMyLocation()
            ActionType.ShowMap -> mControlActionHandler.showMap()
            else -> Unit // Unreachable: caller only dispatches CONTROL_ACTIONS here.
        }
    }

    private fun processVoiceAction(actionType: ActionType) {
        when (actionType) {
            ActionType.DistanceToDestination -> mVoiceActionHandler.speakDistanceToDestination()
            ActionType.DistanceToNextTurn -> mVoiceActionHandler.speakDistanceToNextTurn()
            ActionType.Eta -> mVoiceActionHandler.speakEta()
            ActionType.Mute -> mVoiceActionHandler.processVoiceGuidanceAction(false)
            ActionType.QueryCurrentRoad -> mVoiceActionHandler.speakCurrentRoad()
            ActionType.QueryDestination -> mVoiceActionHandler.speakDestination()
            ActionType.QueryNextTurn -> mVoiceActionHandler.speakNextTurn()
            ActionType.TimeToDestination -> mVoiceActionHandler.speakTimeToDestination()
            ActionType.TimeToNextTurn -> mVoiceActionHandler.speakTimeToNextTurn()
            ActionType.Unmute -> mVoiceActionHandler.processVoiceGuidanceAction(true)
            ActionType.TrafficReport -> mVoiceActionHandler.speakTrafficReport()
            else -> Unit // Unreachable: caller only dispatches VOICE_ACTIONS here.
        }
    }

    private fun processReportAction(actionIntent: GeoActionIntent) {
        when (actionIntent.actionType) {
            ActionType.ReportCrash -> mReportActionHandler.reportCrash(
                actionIntent.accidentType,
                actionIntent.roadDirection,
            )

            ActionType.ReportHazard -> mReportActionHandler.reportHazard(
                actionIntent.hazardType,
                actionIntent.locationOnRoad,
                actionIntent.roadDirection,
            )

            ActionType.ReportPolice -> mReportActionHandler.reportPolice(actionIntent.roadDirection)

            ActionType.ReportRoadClosure -> mReportActionHandler.reportRoadClosure(actionIntent.roadClosureType)

            ActionType.ReportTraffic -> mReportActionHandler.reportTraffic(
                actionIntent.trafficType,
                actionIntent.roadDirection,
            )

            else -> Unit // Unreachable: caller only dispatches REPORT_ACTIONS here.
        }
    }

    private fun processSearchAction(actionIntent: GeoActionIntent) {
        when (actionIntent.actionType) {
            ActionType.ClearSearchResults -> mSearchActionHandler.clearSearchResults()

            ActionType.SelectSearchResult -> mSearchActionHandler.selectSearchResult(actionIntent.searchId)

            ActionType.ApplyElectricVehicleConnectorFilter ->
                mSearchActionHandler.enableElectricVehicleConnectorFilter(true)

            ActionType.RemoveElectricVehicleConnectorFilter ->
                mSearchActionHandler.enableElectricVehicleConnectorFilter(false)

            ActionType.ApplyElectricVehiclePaymentFilter ->
                mSearchActionHandler.enableElectricVehiclePaymentFilter(true)

            ActionType.RemoveElectricVehiclePaymentFilter ->
                mSearchActionHandler.enableElectricVehiclePaymentFilter(false)

            ActionType.ApplyElectricVehicleFastChargingFilter ->
                mSearchActionHandler.enableElectricVehicleFastChargingFilter(true)

            ActionType.RemoveElectricVehicleFastChargingFilter ->
                mSearchActionHandler.enableElectricVehicleFastChargingFilter(false)

            else -> Unit // Unreachable: caller only dispatches SEARCH_ACTIONS here.
        }
    }

    companion object {
        private val TAG: String = GeoIntentProcessor::class.java.simpleName

        private val AVOID_ACTIONS = setOf(
            ActionType.AllowFerries,
            ActionType.AllowHighways,
            ActionType.AllowTolls,
            ActionType.AvoidFerries,
            ActionType.AvoidHighways,
            ActionType.AvoidTolls,
        )

        private val CONTROL_ACTIONS = setOf(
            ActionType.ExitNavigation,
            ActionType.FollowMode,
            ActionType.GoBack,
            ActionType.HideSatellite,
            ActionType.HideTraffic,
            ActionType.RouteOverview,
            ActionType.ShowAlternativeRoutes,
            ActionType.ShowAlternates,
            ActionType.ShowDirectionsList,
            ActionType.ShowSatellite,
            ActionType.ShowTraffic,
            ActionType.MyLocation,
            ActionType.ShowMap,
        )

        private val VOICE_ACTIONS = setOf(
            ActionType.DistanceToDestination,
            ActionType.DistanceToNextTurn,
            ActionType.Eta,
            ActionType.Mute,
            ActionType.QueryCurrentRoad,
            ActionType.QueryDestination,
            ActionType.QueryNextTurn,
            ActionType.TimeToDestination,
            ActionType.TimeToNextTurn,
            ActionType.Unmute,
            ActionType.TrafficReport,
        )

        private val REPORT_ACTIONS = setOf(
            ActionType.ReportCrash,
            ActionType.ReportHazard,
            ActionType.ReportPolice,
            ActionType.ReportRoadClosure,
            ActionType.ReportTraffic,
        )

        private val SEARCH_ACTIONS = setOf(
            ActionType.ClearSearchResults,
            ActionType.SelectSearchResult,
            ActionType.ApplyElectricVehicleConnectorFilter,
            ActionType.RemoveElectricVehicleConnectorFilter,
            ActionType.ApplyElectricVehiclePaymentFilter,
            ActionType.RemoveElectricVehiclePaymentFilter,
            ActionType.ApplyElectricVehicleFastChargingFilter,
            ActionType.RemoveElectricVehicleFastChargingFilter,
        )
    }
}
